rfcs/default_idp_configurable/default_idp_configurable.md

```diff
@@ -0,0 +1,393 @@
+    "title": "Cognito Native Pool",
+    "type": "cognito_user_pool",
+    "credentialHandlingType": "redirect",
+    "signInUri":
+    "signOutUri": 
+    "enableNativeUserPoolUsers": true
+    ...
+  },
+  {
+    ```
+    "title": "Auth0",
+    "type": "cognito_user_pool_federated_idp",
+    "credentialHandlingType": "redirect",
+    "signInUri": 
+    "signOutUri":
+    ...
+  }
+]
+```
+
+- Get id_token. 
+The Service Workbench use **response_type=code** way to integrate with Cognito User Pool, browser redirect to Cognito User Pool login page according to above **Providers Config**. After user input correct username and password, browser redirect user back to Service Workbench page, then request id_token from Service Workbench server-side lambda function(ApiHandler). The lambda function(ApiHandler) send request withing **code** to Cognito user Pool server for id_token.
+
+- Verify id_token.
+API Gateway handles all private request and verify if the token is valid through the lambda function(AuthenticationLayerHandler). The lambda function(AuthenticationLayerHandler) get "tokenValidatorLocator" from "Auth Provider Config", then invoke validateToken method to verify input token.
+
+If user choose federated IDP to authenticate user, the main sequence is the same with Cognito User Pool, there is only a little different for user logining and getting id_token. The callflow like below:
+
+![amazon-cognito-user-pools-saml-idp-auth-flow](./amazon-cognito-user-pools-saml-idp-auth-flow.png)
+
+The user is also redirected to Cognito User Pool Server, but the URL contains **idp_identifier=idpId", Cognito User Pool will go on redirecting to IdP.
+
+# Design Summary
+The general approach proposed by this RFC is to add some configurations to allow Service Workbench to integrate with Keycloak. This will make Service Workbench available in the regions where Cognito User Pool is not available.
+
+## Detail Design
+The design will be divided into two parts: Deployment Process and Runtime Process.
+
+### Deployment Process design
+We keep the current highlevel design that invoking lambda function(postDeployment) to create the resources related with the authentication feature. But which resources are created is determined by configuration as below in "main/config/settings/${STAGE_NAME}.yml" file:
+```
+defaultIDPType: keycloak/cognito
+```
+If user choose "keycloak", because keycloak resources are not created easily during the Service Workbench deployment process, there are some prerequisites that user need to do:
+1. User should have a deployed keycloak.
+2. Create a realm for Service Workbench or use existed realm.
+3. Create a client for Service Workbench.
+4. Create root user in Keycloak.
+
+We will add following configurations options for keycloak integration.
+```
+keyCloakRealm: xxx
+keyCloakClientId: xxx
+keyCloakUrl: xxx
+rootUserEmail: xxx
+rootUserFirstName: xxx
+rootUserLastName: xxx
+```
+![deployment_keycloak](./swb_auth_deploy_keycloak.drawio.png)
```

Comment:
I think the `AddAuthProviders` has two implementations, using the expected one depends on the IdP type configuration. Should polish the architect diagram to reflect the implementation.

Reply:
Yes, I agree, I will refactor the doc.